js/node_shared/unix.js

```diff
@@ -0,0 +1,369 @@
+
+    // associate parent to new child
+    if (globalThis.pidToChildrenTable.has(current_pid)) {
+      globalThis.pidToChildrenTable.set(
+        current_pid,
+        globalThis.pidToChildrenTable.get(current_pid).concat([child])
+      );
+    } else {
+      globalThis.pidToChildrenTable.set(current_pid, [child]);
+    }
+
+    return child.pid;
+  }
+}
+
+//Provides: unix_waitpid
+function unix_waitpid(flags, pid_req) {
+  console.error("unix_waitpid");
+  console.error(flags);
+  console.error(pid_req);
+
+  // TODO: flags stuff is not implemented
+  // TODO: WIFSTOPPED is not currently possible, because I don't know how to
+  // tell if a Javascript process is stopped or killed
+
+  function handleChild(child) {
+    // if it received a signal and is no longer running
+    // let's assume it was killed by the signal
+    let pid = child.pid;
+
+    if (child.status === null) {
+      // WSIGNALED of int
+      let status = [1, globalThis.os.constants.signals[child.signal]];
+      return [0, pid, status];
+    } else {
+      // WEXITED of int
+      let status = [0, child.status];
+      return [0, pid, status];
+    }
+  }
+
+  // any child process
+  if (pid_req == -1) {
+    let children = globalThis.pidToChildrenTable.get(process.pid);
+    if (!children) {
+      throw new Error(`unix_waitpid: no children process of parent {pid_req}`);
+    }
+    for (const child of children) {
+      return handleChild(child);
+    }
+    // specific process
+  } else if (pid_req > 0) {
+    let child = globalThis.pidToProcessTable.get(pid_req);
+    if (!child) {
+      throw new Error(`unix_waitpid: no child at pid {pid_req}`);
+    }
+
+    return handleChild(child);
+    // process group nonsense
+  } else {
+    function getAllDescendants(pid) {
+      let children = globalThis.pidToChildrenTable.get(pid);
+      if (!children) {
+        return [pid];
+      }
+
+      let acc = [];
+      for (const child of children) {
+        acc = acc.concat(getAllDescendants(child.pid));
+      }
+      return [pid].concat(acc);
+    }
+
+    let descendants = getAllDescendants(process.pid);
+    for (const child of descendants) {
+      return handleChild(child);
+    }
+  }
+}
+
+//Provides: unix_pipe
+function unix_pipe(cloexec, vunit) {
+  console.error("unix_pipe");
+  console.error(cloexec);
+  console.error(vunit);
+
+  // The pipe we spawn here will eventually reach unix_spawn.
+  // We don't want to use a real file descriptor, bceause we're trying
+  // to avoid them, in favor of Node's redirection ability, which will open
+  // the corresponding file descriptors for us.
+  // Because this call requires us to decide on what the file descriptor is
+  // ahead of time, though, we will just pass around a simulated FD, which
+  // in conjunction with fdTable, will let us recover the content which is
+  // supposed to be written to that FD.
+  let fd = globalThis.fdCount;
+
+  globalThis.fdCount++;
+
+  return [0, fd, fd];
```

Comment:
I'm not sure what node uses under the hood for polling, but if it's select/poll, this may cause a crash if fdCount > 1024. I'm making the assumption node isn't that bad though, but still commenting for awareness